src/Sasl/Plain.php

```diff
@@ -0,0 +1,104 @@
+<?php
+/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4 foldmethod=marker: */
+// +---------------------------------------------------------------------------
+// | SWAN [ $_SWANBR_SLOGAN_$ ]
+// +---------------------------------------------------------------------------
+// | Copyright $_SWANBR_COPYRIGHT_$
+// +---------------------------------------------------------------------------
+// | Version  $_SWANBR_VERSION_$
+// +---------------------------------------------------------------------------
+// | Licensed ( $_SWANBR_LICENSED_URL_$ )
+// +---------------------------------------------------------------------------
+// | $_SWANBR_WEB_DOMAIN_$
+// +---------------------------------------------------------------------------
+
+namespace Kafka\Sasl;
+
+use Kafka\CommonSocket;
+use Kafka\SaslMechanism;
+use Kafka\Exception;
+
+/**
++------------------------------------------------------------------------------
+* Kafka sasl provider for plain mechanism
++------------------------------------------------------------------------------
+*
+* @package
+* @version $_SWANBR_VERSION_$
+* @copyright Copyleft
+* @author $_SWANBR_AUTHOR_$
++------------------------------------------------------------------------------
+*/
+
+class Plain extends Mechanism implements SaslMechanism
+{
+    // {{{ consts
+    
+    private const MECHANISM_NAME = "PLAIN";
+
+    // }}}
+    // {{{ members
+
+    private $username;
+
+    private $password;
+
+    // }}}
+    // {{{ functions
+    // {{{ public function __construct()
+    
+    /**
+     *
+     * __construct
+     *
+     * @access public
+     * @return void
+     */
+    public function __construct(array $options)
+    {
+        if (! isset($options['username']) || ! isset($options['password'])) {
+            throw new Exception('Sasl plain given options is invalid, must give username and password');
+        }
+
+        $this->username = trim($options['username']);
+        $this->password = trim($options['password']);
+    }
+
+    // }}}
+    // {{{ public function authenticate()
+    
+    /**
+     *
+     * sasl authenticate
+     *
+     * @access public
+     * @return void
+     */
+    public function authenticate(CommonSocket $socket)
```

Comment:
It feels like this method shouldn't be here but rather in the base class, something like:

```php
public function authenticate(CommonSocket $socket): void
{
    $this->handShake($socket, $this->getMechanismName());
   
    $socket->selectWrite($this->createAuthenticationMessage());
    $socket->selectRead(4, true);
}

abstract protected function createAuthenticationMessage(): string;
```

Reply:
GSSAPI and SCRAM mechanisms require interactive multiple messages that are different from the server side， so I think needless abstract `createAuthenticationMessage`